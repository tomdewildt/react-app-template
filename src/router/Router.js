import React from "react";
import {
    BrowserRouter,
    Redirect,
    Route,
    Switch,
} from "react-router-dom";
import PropTypes from "prop-types";

import { PageHome } from "../pages";

const Router = ( { children } ) => (
    <BrowserRouter>
        { children }
        <Switch>
            <Route exact path="/" component={ PageHome } />
            <Route exact path="*" component={ () => <Redirect to="/" /> } />
        </Switch>
    </BrowserRouter>
);

Router.propTypes = {
    children: PropTypes.node,
};

Router.defaultProps = {
    children: null,
};

export default Router;
