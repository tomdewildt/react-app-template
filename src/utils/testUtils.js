/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import React from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { render } from "@testing-library/react";
import { expectSaga } from "redux-saga-test-plan";
import createStore from "redux-mock-store";

import defaultTheme from "../theme";

const defaultStore = createStore( [] );
const defaultState = { map: {} };

export const renderWithRouter = (
    component,
    entries = [ "/" ],
    state = defaultState,
    theme = defaultTheme,
) => {
    const store = defaultStore( state );
    return {
        store,
        ...render( component, {
            wrapper: ( { children } ) => (
                <ThemeProvider theme={ theme }>
                    <Provider store={ store }>
                        <MemoryRouter initialEntries={ entries }>
                            { children }
                        </MemoryRouter>
                    </Provider>
                </ThemeProvider>
            ),
        } ),
    };
};

export const renderWithProvider = (
    component,
    state = defaultState,
    theme = defaultTheme,
) => {
    const store = defaultStore( state );
    return {
        store,
        ...render( component, {
            wrapper: ( { children } ) => (
                <ThemeProvider theme={ theme }>
                    <Provider store={ store }>
                        { children }
                    </Provider>
                </ThemeProvider>
            ),
        } ),
    };
};

export const renderWithTheme = (
    component,
    theme = defaultTheme,
) => render( component, {
    wrapper: ( { children } ) => (
        <ThemeProvider theme={ theme }>
            { children }
        </ThemeProvider>
    ),
} );

export const runReducer = ( reducer, action, state ) => reducer( state, action );

export const runSaga = ( saga, ...args ) => expectSaga( saga, ...args );
