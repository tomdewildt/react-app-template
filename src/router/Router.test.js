import React from "react";
import { axe } from "jest-axe";

import { renderWithRouter } from "../utils/testUtils";

import Router from "./Router";

describe( "Router", () => {
    it( "renders consistently", () => {
        const { asFragment } = renderWithRouter( <Router /> );
        const fragment = asFragment();

        expect( fragment ).toMatchSnapshot();
    } );

    it( "should have no axe violations", async () => {
        const { container } = renderWithRouter( <Router /> );
        const result = await axe( container );

        expect( result ).toHaveNoViolations();
    } );

    it( "renders home page", () => {
        const { container } = renderWithRouter( <Router />, [ "/" ] );

        expect( container.querySelector( "#home" ) ).toBeInTheDocument();
    } );

    it( "redirects to home page", () => {
        const { container } = renderWithRouter( <Router />, [ "/test" ] );

        expect( container.querySelector( "#home" ) ).toBeInTheDocument();
    } );
} );
