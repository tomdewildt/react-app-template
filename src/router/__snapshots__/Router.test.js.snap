// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router renders consistently 1`] = `
<DocumentFragment>
  <main
    class=""
    id="home"
  />
</DocumentFragment>
`;
